ast/index.go

```diff
@@ -246,15 +246,20 @@ func (i *refindices) Update(rule *Rule, expr *Expr) {
 
 	op := expr.Operator()
 
-	if op.Equal(Equality.Ref()) {
+	switch {
+	case op.Equal(Equality.Ref()):
 		i.updateEq(rule, expr)
-	} else if op.Equal(Equal.Ref()) && len(expr.Operands()) == 2 {
+
+	case op.Equal(Equal.Ref()) && len(expr.Operands()) == 2:
 		// NOTE(tsandall): if equal() is called with more than two arguments the
 		// output value is being captured in which case the indexer cannot
 		// exclude the rule if the equal() call would return false (because the
 		// false value must still be produced.)
 		i.updateEq(rule, expr)
-	} else if op.Equal(GlobMatch.Ref()) {
+
+	case op.Equal(GlobMatch.Ref()) && len(expr.Operands()) == 3:
```

Comment:
Maybe this is spurious on my part, but I see that this case isn't quite an exact match to the `else if` check it's replacing. Are there any cases where `op.Equal(GlobMatch.Ref())` could occur *without* having 3 items in `expr.Operands()`?

If that *never* happens, then all is well. I'd hope that our checks further up the chain ensure there's at least 3x operands here. :sweat_smile:

Reply:
> Are there any cases where op.Equal(GlobMatch.Ref()) could occur without having 3 items in expr.Operands()?

Yeah, that's when the output is captured (see the test case additions). We _don't_ want to use the index for statements like this. The logic here is that when something contributes to the index, it'll contribute to _excluding rules_ for evaluation. So if we skip it here, it'll mean the rule body is still returned. Hope this makes sense 😅

Reply:
Perfect! So we're _intentionally excluding_ the other cases, because the original logic was too broad in what it'd pick up for indexing, and only the 3x operand case is useful to index for glob matching.

That answers my question then. :+1: